packages/next/src/client/components/router-reducer/reducers/server-reference-info.ts

```diff
@@ -0,0 +1,77 @@
+export interface ServerReferenceInfo {
+  type: 'server-action' | 'use-cache'
+  usedArgs: [boolean, boolean, boolean, boolean, boolean, boolean]
+  hasRestArgs: boolean
+}
+
+/**
+ * Extracts info about the server reference for the given server reference ID by
+ * parsing the first byte of the hex-encoded ID.
+ *
+ * ```
+ * Bit positions: [7]      [6] [5] [4] [3] [2] [1]  [0]
+ * Bits:          typeBit  argMask                  restArgs
+ * ```
+ *
+ * If the `typeBit` is `1` the server reference represents a `"use cache"`
+ * function, otherwise a server action.
+ *
+ * The `argMask` encodes whether the function uses the argument at the
+ * respective position.
+ *
+ * The `restArgs` bit indicates whether the function uses a rest parameter.
+ *
+ * @param id hex-encoded server reference ID
+ */
+export function extractInfoFromServerReferenceId(
+  id: string
+): ServerReferenceInfo {
+  // Bit positions: [7]      [6] [5] [4] [3] [2] [1]  [0]
+  // Bits:          typeBit  argMask                  restArgs
+  const infoByte = parseInt(id.slice(0, 2), 16)
+  const typeBit = (infoByte >> 7) & 0x1
+  const argMask = (infoByte >> 1) & 0x3f
+  const restArgs = infoByte & 0x1
+  const usedArgs = Array(6)
+
+  for (let index = 0; index < 6; index++) {
+    const bitPosition = 5 - index
+    const bit = (argMask >> bitPosition) & 0x1
+    usedArgs[index] = bit === 1
+  }
+
+  return {
+    type: typeBit === 1 ? 'use-cache' : 'server-action',
+    usedArgs: usedArgs as [
+      boolean,
+      boolean,
+      boolean,
+      boolean,
+      boolean,
+      boolean,
+    ],
+    hasRestArgs: restArgs === 1,
+  }
+}
+
+/**
+ * Creates a sparse array containing only the used arguments based on the
+ * provided action info.
+ */
+export function filterArgs(
+  args: unknown[],
+  info: ServerReferenceInfo
+): unknown[] {
+  const filteredArgs = new Array(args.length)
```

Comment:
is there a reason why we want a sparse array vs just filling the unused positions with `undefined` or something?

Reply:
I thought it matched the concept of omitting args better. But practically it doesn't really matter because in the Flight payload there is no difference between a sparse array and an array with `undefined` items.

Reply:
i guess this doesn't matter much, i just reflexively avoid holey arrays bc of map/filter weirdness